{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "%reset\n",
    "\n",
    "import pandas as pd\n",
    "import argparse\n",
    "import numpy as np\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mm10: 23006 experiments\n",
      "mm10: 0.6816917325914978 passed QC\n",
      "True\n",
      "True\n",
      "Grip1\n",
      "Grip1\n",
      "Grip1\n",
      "Grip1\n",
      "Grip1\n",
      "Grip1\n",
      "Wdr11\n",
      "Wdr11\n",
      "Wdr11\n",
      "Htt\n",
      "Htt\n",
      "Htt\n",
      "Htt\n",
      "Htt\n",
      "Htt\n",
      "Htt\n",
      "Mpnd\n",
      "Mpnd\n",
      "Htt\n",
      "Htt\n",
      "mm10: 0.8666071542434484 in TF list\n",
      "mm10: 712 unique TF\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "version = 'v3'\n",
    "genome = 'mm10'\n",
    "infile_chip = f\"../resources/experimentList_{version}.tab\"\n",
    "infile_tfs = f\"../resources/{genome}/TF_list.txt\"\n",
    "geneid_genename_synonym_table = f\"../genome/{genome}/{genome}_ENSID_Genename_synonyms.txt.gz\"\n",
    "synonym_genename_dict = f\"../genome/{genome}/GeneName_Synonyms_dict.pkl\"\n",
    "th_reads = 1000\n",
    "th_mapped_reads = 0.1\n",
    "th_duplicates = .95\n",
    "th_peaks = 1\n",
    "th_exp_per_tf = 128\n",
    "\n",
    "# get experiment table\n",
    "chip = pd.read_csv(infile_chip,sep='\\t',header=None,usecols=[0,1,2,3,4,5,6,7,8],index_col=0)\n",
    "chip.columns = ['genome','antigen_class','antigen','celltype_class','celltype','celltype_description','QC','title']\n",
    "\n",
    "# get only TFs and others from genome\n",
    "chip = chip[ (chip.genome == genome) & (chip.antigen_class=='TFs and others') ]\n",
    "\n",
    "# parse QC column add to chip table\n",
    "QC = pd.DataFrame([[float(n) for n in qc.split(',')] for qc in chip.QC],columns=['n_reads','f_mapped','f_duplicates','n_peaks'],index=chip.index)\n",
    "QC.iloc[:,1] /= 100\n",
    "QC.iloc[:,2] /= 100\n",
    "chip = pd.concat([chip,QC],axis=1)\n",
    "\n",
    "# get peaks per unique mapped reads\n",
    "chip.loc[:,'n_peaks_per_unique_mapped_reads'] = chip.n_peaks/(chip.f_mapped*chip.n_reads*(1-chip.f_duplicates))\n",
    "n_tot = chip.shape[0]\n",
    "print(f'{genome}: {n_tot} experiments')\n",
    "\n",
    "# apply thresholds\n",
    "idx_out = list( chip[(chip['n_reads']     < th_reads) | \n",
    "                     (chip['f_mapped']    < th_mapped_reads) |\n",
    "                     (chip['f_duplicates']> th_duplicates) |\n",
    "                     (chip['n_peaks']     < th_peaks) ].index )\n",
    "\n",
    "# For TFs with more than th experiments, keep the th with the highest n_peaks_per_unique_mapped_reads\n",
    "exp_per_tf = chip.groupby('antigen')['antigen'].aggregate('count')\n",
    "for tf in exp_per_tf.loc[exp_per_tf > th_exp_per_tf].index:\n",
    "    idx_out.extend( list( chip.loc[chip.antigen==tf].sort_values('n_peaks_per_unique_mapped_reads',ascending=False).index[th_exp_per_tf:] ) )\n",
    "chip.drop(idx_out,inplace=True)\n",
    "\n",
    "print(f'{genome}: {chip.shape[0]/n_tot} passed QC')\n",
    "\n",
    "print( 'Ar' in np.sort(chip.antigen) )\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "# load GeneID GeneName Synonym table gene name list\n",
    "Gene_id_name_syn = pd.read_csv(geneid_genename_synonym_table,sep='\\t')\n",
    "GeneName = set(Gene_id_name_syn['Gene name'])\n",
    "\n",
    "Gene_id_name_syn.replace({np.nan:'None'},inplace=True)\n",
    "Gene_id_name_syn.drop(Gene_id_name_syn.index[Gene_id_name_syn['Gene Synonym']==\"None\"],inplace=True)\n",
    "Gene_id_name_syn.drop_duplicates(inplace=True)\n",
    "Gene_id_name_syn.drop( Gene_id_name_syn.loc[Gene_id_name_syn['Gene Synonym'] == Gene_id_name_syn['Gene name']].index, inplace=True)\n",
    "\n",
    "print( 'Ar' in GeneName )\n",
    "\n",
    "\n",
    "# load synonym to gene name dictionary\n",
    "with open(synonym_genename_dict, 'rb') as f:\n",
    "    Synonym_2_GeneName = pickle.load(f)\n",
    "\n",
    "# get antigen gene names to change\n",
    "to_rename = []\n",
    "not_found = []\n",
    "for g in chip.antigen.unique():\n",
    "    if g in GeneName:\n",
    "        continue\n",
    "    else:\n",
    "        if g in Synonym_2_GeneName.keys():\n",
    "            to_rename.append([g,Synonym_2_GeneName[g]])\n",
    "        else:\n",
    "            not_found.append(g)\n",
    "to_rename = dict(zip(np.array(to_rename)[:,0],np.array(to_rename)[:,1]))\n",
    "\n",
    "# change gene names\n",
    "for g in to_rename.keys():\n",
    "    idx = chip[ chip.antigen==g ].index\n",
    "    for i in idx:\n",
    "        chip.at[i,'antigen'] = to_rename[chip.at[i,'antigen']]\n",
    "\n",
    "# get TF list\n",
    "with open(infile_tfs,'r') as f:\n",
    "    TFs = [tf.strip() for tf in f.readlines()]\n",
    "TFs = np.array(TFs)\n",
    "\n",
    "# get tf names to change\n",
    "to_rename = []\n",
    "not_found = []\n",
    "for g in TFs:\n",
    "    if g in GeneName:\n",
    "        continue\n",
    "    else:\n",
    "        if g in Synonym_2_GeneName.keys():\n",
    "            to_rename.append([g,Synonym_2_GeneName[g]])\n",
    "        else:\n",
    "            not_found.append(g)\n",
    "to_rename = dict(zip(np.array(to_rename)[:,0],np.array(to_rename)[:,1]))\n",
    "\n",
    "# change tf names\n",
    "for g in to_rename.keys():\n",
    "    i = np.where( TFs == g )[0]\n",
    "    TFs[i] = to_rename[g]\n",
    "\n",
    "\n",
    "# Keep only antigens that are in TF list\n",
    "tf_id = []\n",
    "tf_out = []\n",
    "for id in chip.index:\n",
    "    antigen = chip.at[id,'antigen']\n",
    "    # if antigen is in TF list add id\n",
    "    if antigen in TFs:\n",
    "        tf_id.append(id)\n",
    "    # if antigen is a synonym of \n",
    "    elif antigen in Gene_id_name_syn['Gene Synonym'].values:\n",
    "        print(antigen)\n",
    "        gene_name = Gene_id_name_syn.loc[ Gene_id_name_syn['Gene Synonym']==antigen, 'Gene name'].values\n",
    "        if any([g in TFs for g in gene_name]):\n",
    "            tf_id.append(id)\n",
    "    else:\n",
    "        tf_out.append(antigen)\n",
    "\n",
    "# print kept ratio\n",
    "print(f'{genome}: {len(tf_id)/chip.shape[0]} in TF list')\n",
    "chip = chip.loc[tf_id,:]\n",
    "print(f'{genome}: {len(chip.antigen.unique())} unique TF')\n",
    "\n",
    "print( 'Ar' in np.sort(chip.antigen) )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "chip",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
